### PR TITLE
Build Mac wheels in CI

### PR DESCRIPTION
**Pull request by scotts**
Note that this summary was written a month after opening the PR. As evidenced by the commit history below, getting this in shape has been a journey. A summary of the actual changes we'll commit:

1. A separate workflow for compiling and testing Mac wheels. While it has a lot of overlap with the workflow for Linux wheels, there is still some specialization needed when building the wheel. In particular, we are actually still using the [test-infra@macbuildwheel](https://github.com/pytorch/test-infra/tree/macbuildwheel) branch. Eventually that branch will get merged, and we can use main, just like with Linux. At that point, it will be worth seeing if we can consolidate the workflows and make Mac vs Linux a matrix option. (And maybe CUDA as well, see #272.)
2. Minor changes to the Linux wheel workflow to keep it aligned with Mac wheels.
3. Generalization of some of our packaging scripts to be aware of "Linux" versus "Darwin."
4. The removal of logging calls (`VLOG` and `LOG`) from our C++ code. For reasons that we have not been able to fully explain, we get segfaults in those logging calls only when we ran from a Mac wheel. We do not segfault there when we build and test Mac from source, or in any situation with Linux. That might indicate a problem in PyTorch core. If we want logging back, we can decide how best to it then.

### CODE REVIEW

File: .github/workflows/macos_wheel.yaml
Comment (atalman) [inline .github/workflows/macos_wheel.yaml:28]:
```suggestion
    uses: pytorch/test-infra/.github/workflows/generate_binary_build_matrix.yml@macbuildwheel
```

File: .github/workflows/macos_wheel.yaml
Comment (atalman) [inline .github/workflows/macos_wheel.yaml:42]:
```suggestion
    uses: pytorch/test-infra/.github/workflows/build_wheels_macos.yml@macbuildwheel
```

File: .github/workflows/macos_wheel.yaml
Comment (atalman) [inline .github/workflows/macos_wheel.yaml:71]: Looks like this step is not working as intended. I get something like this:
```
....
  # To activate this environment, use
  #
  #     $ conda activate test
  #
  # To deactivate an active environment, use
  #
  #     $ conda deactivate
  

setup-miniconda ran successfully
```

However next step:
```
Run python3 -m pip install --upgrade pip
  python3 -m pip install --upgrade pip
  shell: /bin/bash -l -eo pipefail {0}
  env:
    INPUT_RUN_POST: true
    CONDA: /Users/ec[2](https://github.com/pytorch/torchcodec/actions/runs/11260601615/job/31315172014?pr=230#step:4:2)-user/miniconda3
    CONDA_PKGS_DIR: /Users/ec2-user/conda_pkgs_dir
Requirement already satisfied: pip in /Library/Frameworks/Python.framework/Versions/[3](https://github.com/pytorch/torchcodec/actions/runs/11260601615/job/31315172014?pr=230#step:4:3).10/lib/python3.10/site-packages (24.2)
```

It picks python 3.10

File: .github/workflows/macos_wheel.yaml
Comment (atalman) [inline .github/workflows/macos_wheel.yaml:103]: For this step I suggest we onboard to validation framework:
https://github.com/pytorch/test-infra/blob/main/.github/workflows/validate-domain-library.yml

This is exactly the use case we wrote validation framework for:
https://github.com/pytorch/test-infra/wiki/Validation-Framework

File: .github/workflows/macos_wheel.yaml
Comment (atalman) [inline .github/workflows/macos_wheel.yaml:164]: Here all conda install need to be done via ${CONDA_RUN}

**Comment by atalman** [comment]
Hi @scotts 
Looks like problem with the build 
```
pip install TorchCodec-0.0.4.dev0-cp39-cp39-macosx_11_0_arm64.whl
pip install --pre torchvision --index-url https://download.pytorch.org/whl/nightly/cpu
conda install "ffmpeg=4.4.2" -c conda-forge
```
and then executing smoke tests:
```
python test/decoders/manual_smoke_test.py  
zsh: segmentation fault  python test/decoders/manual_smoke_test.py
```

File: .github/workflows/linux_wheel.yaml
Comment (scotts) [inline .github/workflows/linux_wheel.yaml:38]: I learned about this option from @ahmadsharif1. I turned it on for both Linux and Mac, as I think it fits our use-case where we're also building C++ libraries.

File: .github/workflows/linux_wheel.yaml
Comment (scotts) [inline .github/workflows/linux_wheel.yaml:126]: In case we run into any problems again on Linux or Mac, I think running the smoke test with the faulthandler on will be helpful.

File: .github/workflows/macos_wheel.yaml
Comment (ahmadsharif1) [inline .github/workflows/macos_wheel.yaml:48]: Does this build the wheel against nightly?

I noticed we are testing against nightly so probably have to build against nightly too?

File: .github/workflows/macos_wheel.yaml
Comment (NicolasHug) [inline .github/workflows/macos_wheel.yaml:103]: Nit: either copy the comment made in the linux file https://github.com/NicolasHug/torchcodec/blob/d43dd9120d227ad922079830c05b3c8fa3662de4/.github/workflows/linux_wheel.yaml#L110-L119 or just point to it, to explain why we do this?

Once we merge this PR and when we have the CUDA wheel builds, it'll be time to refactor the 3 separate scripts into something more centralized

File: .github/workflows/macos_wheel.yaml
Comment (NicolasHug) [reply .github/workflows/macos_wheel.yaml:48]: Yes, I think the `build_wheels_macos` job of test-infra is setup to build against nightly by default, and we can see from the logs that nightly is pickedup

**Review by NicolasHug — APPROVED** [review]
What a ride. Thank you so much @scotts for pushing through!
Nothing blocking from me, just nits and one Q

File: .github/workflows/macos_wheel.yaml
Comment (scotts) [reply .github/workflows/macos_wheel.yaml:58]: No, it shouldn't still be needed. Let me remove and check. If it is still needed, I'll make a comment.

**Review by ahmadsharif1 — APPROVED** [review]

File: .github/workflows/macos_wheel.yaml
Comment (scotts) [reply .github/workflows/macos_wheel.yaml:48]: I don't know, but this code is common between Mac and Linux, and was inherited from Linux. Let's investigate that as a follow-up.

**Pull request merged**
